File: X2/bernoulli.php
<?php
// Progp X2 Bernoulli
// Authors: Johan Callvik & Mauritz Zachrisson

$input = $argv[1];

function B($n) {
	$B = array();
	$B[0] = 1;

	for ($m = 1; $m <= $n; $m++) { 
		$B[$m] = 0;
		for ($k = 0; $k <= $m - 1; $k++) { 
			$B[$m] = $B[$m] - binom($m + 1, $k) * $B[$k];
		}
		$B[$m] = $B[$m] / ($m + 1);
	}

	return $B[$n];
}

function binom($n, $k) {
	$r = 1;
	for ($i = 1; $i <= $k; $i++) {
		$r = $r * ($n - $i + 1) / $i;
	}
	return $r;
}

echo B($input) . "\n";

?>

File: X2/bernoulli.clj
;; Author: Johan Callvik & Mauritz Zachrisson
(ns bernoulli)
 

;; This function calculates the desired binominalcoefficent given two arguments.
(defn binom [n k]
    (loop [i 1 r 1.0]
      (if (>= i (+ k 1))
        r
        (recur (+ i 1) (* r (/ (+ (- n i) 1) i))))))



(defn bern [m]
	(if (= m 0)
		1
		(- 0 	
			(/ (apply + (for [k (range m)] (* (binom (+ m 1) k) (bern k))))		;; for loops through values of k up to m and calculates that bernoulli number.
			(+ m 1)																;; apply sums all the bernoulli numbers which are then divided by m + 1.
		)
	)))

;; Prints the first 20 bernoulli numbers.
(println (map (fn [y] (bern y)) (range 20)))



